File: tests/Bots/refs_utilsTest.php
<?php



use PHPUnit\Framework\TestCase;

use function WpRefs\Bots\RefsUtils\endsWith;
use function WpRefs\Bots\RefsUtils\strstartswith;
use function WpRefs\Bots\RefsUtils\del_start_end;
use function WpRefs\Bots\RefsUtils\remove_start_end_quotes;

class refs_utilsTest extends TestCase
{

    /**
     * @test
     * @description يضيف علامات اقتباس مزدوجة لنص عادي.
     */
    public function test_adds_double_quotes_to_plain_string()
    {
        $this->assertEquals('"value"', remove_start_end_quotes('value'));
    }

    /**
     * @test
     * @description يزيل علامات الاقتباس المفردة ويضيف مزدوجة.
     */
    public function test_replaces_single_quotes_with_double_quotes()
    {
        // سلوك الدالة هو إزالة الاقتباسات الموجودة ثم إضافة جديدة.
        $this->assertEquals('"value"', remove_start_end_quotes("'value'"));
    }

    /**
     * @test
     * @description يزيل علامات الاقتباس المزدوجة ويضيف مزدوجة مرة أخرى.
     */
    public function test_replaces_double_quotes_with_double_quotes()
    {
        $this->assertEquals('"value"', remove_start_end_quotes('"value"'));
    }

    /**
     * @test
     * @description يحيط النص بعلامات اقتباس مفردة إذا كان يحتوي على علامات مزدوجة بالداخل.
     */
    public function test_wraps_with_single_quotes_if_contains_double_quotes()
    {
        $this->assertEquals("'val\"ue'", remove_start_end_quotes('val"ue'));
    }

    /**
     * @test
     * @description يزيل المسافات الزائدة من البداية والنهاية.
     */
    public function test_trims_whitespace()
    {
        $this->assertEquals('"value"', remove_start_end_quotes('  value  '));
    }

    /**
     * @test
     * @description يتعامل مع نص فارغ.
     */
    public function test_handles_empty_string()
    {
        $this->assertEquals('""', remove_start_end_quotes(''));
    }
    public function test_one_quotes_double()
    {
        $this->assertEquals("'\"value'", remove_start_end_quotes('  "value '));
    }
    public function test_one_quotes_single()
    {
        $this->assertEquals('"\'value"', remove_start_end_quotes("  'value "));
    }

    // اختبارات دالة endsWith
    public function testEndsWith()
    {
        $tests = [
            // حالة: ينتهي بالنص المطلوب
            ["string" => "Hello world", "endString" => "world", "expected" => true],
            // حالة: لا ينتهي بالنص المطلوب
            ["string" => "Hello world", "endString" => "hello", "expected" => false],
            // حالة: نص فارغ
            ["string" => "", "endString" => "test", "expected" => false],
            // حالة: نص البحث فارغ
            // ["string" => "test", "endString" => "", "expected" => true],
            // حالة: نص البحث أطول من النص الأصلي
            ["string" => "short", "endString" => "longer text", "expected" => false],
            // حالة: تطابق كامل
            ["string" => "exact", "endString" => "exact", "expected" => true],
            // حالة: أحرف خاصة
            ["string" => "file.txt", "endString" => ".txt", "expected" => true],
            // حالة: حساسية الأحرف
            ["string" => "Case", "endString" => "case", "expected" => false]
        ];

        foreach ($tests as $test) {
            $result = endsWith($test['string'], $test['endString']);
            $this->assertEquals($test['expected'], $result, $test['string']);
        }
    }

    // اختبارات دالة strstartswith
    public function testStrstartswith()
    {
        $tests = [
            // حالة: يبدأ بالنص المطلوب
            ["text" => "Hello world", "start" => "Hello", "expected" => true],
            // حالة: لا يبدأ بالنص المطلوب
            ["text" => "Hello world", "start" => "world", "expected" => false],
            // حالة: نص فارغ
            ["text" => "", "start" => "test", "expected" => false],
            // حالة: نص البحث فارغ
            ["text" => "test", "start" => "", "expected" => true],
            // حالة: نص البحث أطول من النص الأصلي
            ["text" => "short", "start" => "longer text", "expected" => false],
            // حالة: تطابق كامل
            ["text" => "exact", "start" => "exact", "expected" => true],
            // حالة: أحرف خاصة
            ["text" => "#tag", "start" => "#", "expected" => true],
            // حالة: حساسية الأحرف
            ["text" => "Case", "start" => "case", "expected" => false]
        ];

        foreach ($tests as $test) {
            $result = strstartswith($test['text'], $test['start']);
            $this->assertEquals($test['expected'], $result);
        }
    }

    // اختبارات دالة del_start_end
    public function testDelStartEnd()
    {
        $tests = [
            // حالة: إزالة من البداية والنهاية
            ["text" => "'quoted text'", "find" => "'", "expected" => "quoted text"],
            // حالة: إزالة علامات تنصيص مزدوجة
            ["text" => '"double quoted"', "find" => '"', "expected" => "double quoted"],
            // حالة: نص بدون علامات في البداية والنهاية
            ["text" => "no quotes", "find" => "'", "expected" => "no quotes"],
            // حالة: علامة في البداية فقط
            ["text" => "'start only", "find" => "'", "expected" => "'start only"],
            // حالة: علامة في النهاية فقط
            ["text" => "end only'", "find" => "'", "expected" => "end only'"],
            // حالة: مسافات زائدة
            // ["text" => "  '  spaced  '  ", "find" => "'", "expected" => "  spaced  "],
            ["text" => "  '  spaced  '  ", "find" => "'", "expected" => "spaced"],
            // حالة: نص فارغ
            ["text" => "", "find" => "'", "expected" => ""],
            // حالة: علامة البحث فارغة
            ["text" => "test", "find" => "", "expected" => "test"],
            // حالة: علامات متعددة
            ["text" => "''multiple''", "find" => "'", "expected" => "'multiple'"],
            // حالة: نص يتكون من العلامة فقط
            ["text" => "''", "find" => "'", "expected" => ""]
        ];

        foreach ($tests as $test) {
            $result = del_start_end($test['text'], $test['find']);
            $this->assertEquals($test['expected'], $result);
        }
    }

    // اختبارات دالة remove_start_end_quotes
    public function testFixAttrValue()
    {
        $tests = [
            // حالة: نص بدون علامات تنصيص
            ["text" => "value1", "expected" => '"value1"'],
            // حالة: نص بعلامات تنصيص مفردة
            ["text" => "'value2'", "expected" => '"value2"'],
            // حالة: نص بعلامات تنصيص مزدوجة
            ["text" => '"value3"', "expected" => '"value3"'],
            // حالة: نص بعلامات تنصيص مختلطة
            ["text" => '"mixed\'quotes"', "expected" => '"mixed\'quotes"'],
            // حالة: نص يحتوي على علامات تنصيص داخلياً
            ["text" => 'value"with"quotes', "expected" => "'value\"with\"quotes'"],
            // حالة: مسافات زائدة
            ["text" => "  spaced  ", "expected" => '"spaced"'],
            // حالة: نص يحتوي على علامات تنصيص في المنتصف
            ["text" => "val'ue", "expected" => '"val\'ue"']
        ];

        foreach ($tests as $test) {
            $result = remove_start_end_quotes($test['text']);
            $this->assertEquals($test['expected'], $result, $test['text']);
        }
    }
    public function testFixEmpty()
    {
        $this->assertEquals("", "");
    }
    public function testFixOnlyQuotes()
    {
        $this->assertEquals('""', remove_start_end_quotes('""'));
    }
    public function testFixOnlySingleQuotes()
    {
        $this->assertEquals('""', remove_start_end_quotes("''"));
    }
}

File: tests/Bots/refs_utilsTest.py
from __future__ import annotations

from tests.conftest import assert_equal_compare
from src.bots.refs_utils import remove_start_end_quotes, rm_str_from_start_and_end


def str_ends_with(string: str, end: str) -> bool:
    return string.endswith(end)


def str_starts_with(string: str, start: str) -> bool:
    return string.startswith(start)


class TestRefsUtils:
    def test_adds_double_quotes_to_plain_string(self) -> None:
        assert remove_start_end_quotes("value") == '"value"'

    def test_replaces_single_quotes_with_double_quotes(self) -> None:
        assert remove_start_end_quotes("'value'") == '"value"'

    def test_replaces_double_quotes_with_double_quotes(self) -> None:
        assert remove_start_end_quotes('"value"') == '"value"'

    def test_wraps_with_single_quotes_if_contains_double_quotes(self) -> None:
        assert remove_start_end_quotes('val"ue') == "'val\"ue'"

    def test_trims_whitespace(self) -> None:
        assert remove_start_end_quotes("  value  ") == '"value"'

    def test_handles_empty_string(self) -> None:
        assert remove_start_end_quotes("") == '""'

    def test_one_quotes_double(self) -> None:
        assert remove_start_end_quotes('  "value ') == "'\"value'"

    def test_one_quotes_single(self) -> None:
        assert remove_start_end_quotes("  'value ") == '"\'value"'

    def test_str_ends_with(self) -> None:
        tests = [
            {"string": "Hello world", "endString": "world", "expected": True},
            {"string": "Hello world", "endString": "hello", "expected": False},
            {"string": "", "endString": "test", "expected": False},
            {"string": "short", "endString": "longer text", "expected": False},
            {"string": "exact", "endString": "exact", "expected": True},
            {"string": "file.txt", "endString": ".txt", "expected": True},
            {"string": "Case", "endString": "case", "expected": False},
        ]

        for case in tests:
            result = str_ends_with(case["string"], case["endString"])
            assert_equal_compare(case["expected"], case["string"], result)

    def test_str_starts_with(self) -> None:
        tests = [
            {"text": "Hello world", "start": "Hello", "expected": True},
            {"text": "Hello world", "start": "world", "expected": False},
            {"text": "", "start": "test", "expected": False},
            {"text": "test", "start": "", "expected": True},
            {"text": "short", "start": "longer text", "expected": False},
            {"text": "exact", "start": "exact", "expected": True},
            {"text": "#tag", "start": "#", "expected": True},
            {"text": "Case", "start": "case", "expected": False},
        ]

        for case in tests:
            result = str_starts_with(case["text"], case["start"])
            assert_equal_compare(case["expected"], case["text"], result)

    def test_del_start_end(self) -> None:
        tests = [
            {"text": "'quoted text'", "find": "'", "expected": "quoted text"},
            {"text": '"double quoted"', "find": '"', "expected": "double quoted"},
            {"text": "no quotes", "find": "'", "expected": "no quotes"},
            {"text": "'start only", "find": "'", "expected": "'start only"},
            {"text": "end only'", "find": "'", "expected": "end only'"},
            {"text": "  '  spaced  '  ", "find": "'", "expected": "spaced"},
            {"text": "", "find": "'", "expected": ""},
            {"text": "''multiple''", "find": "'", "expected": "'multiple'"},
            {"text": "''", "find": "'", "expected": ""},
        ]

        for case in tests:
            result = rm_str_from_start_and_end(case["text"], case["find"])
            assert_equal_compare(case["expected"], case["text"], result)

    def test_remove_start_end_quotes(self) -> None:
        tests = [
            {"text": "value1", "expected": '"value1"'},
            {"text": "'value2'", "expected": '"value2"'},
            {"text": '"value3"', "expected": '"value3"'},
            {"text": '"mixed\'quotes"', "expected": '"mixed\'quotes"'},
            {"text": 'value"with"quotes', "expected": "'value\"with\"quotes'"},
            {"text": "  spaced  ", "expected": '"spaced"'},
            {"text": "val'ue", "expected": '"val\'ue"'},
        ]

        for case in tests:
            result = remove_start_end_quotes(case["text"])
            assert_equal_compare(case["expected"], case["text"], result)

    def test_fix_empty(self) -> None:
        assert "" == ""

    def test_fix_only_quotes(self) -> None:
        assert remove_start_end_quotes('""') == '""'

    def test_fix_only_single_quotes(self) -> None:
        assert remove_start_end_quotes("''") == '""'

    def test_del_start_end_empty(self) -> None:
        result = rm_str_from_start_and_end("testzz", "")
        assert_equal_compare("testzz", "testzz", result)
